dbt/include/databricks/macros/materializations/table.sql

```diff
@@ -42,5 +42,18 @@
 # --- Autogenerated dbt materialization code. --- #
 dbt = dbtObj(spark.table)
 df = model(dbt, spark)
+
+# make sure pandas exists
+import importlib.util
+import pyspark.pandas
+package_name = 'pandas'
+if importlib.util.find_spec(package_name):
+    import pandas
+    # convert to pyspark.sql.dataframe.DataFrame
+    if isinstance(df, pandas.core.frame.DataFrame):
+      df = spark.createDataFrame(df)
+    elif isinstance(df, pyspark.pandas.frame.DataFrame):
+      df = df.to_spark()
```

Comment:
The following flow would be better?

- if df is pandas DataFrame,
  - if pyspark.pandas is available, convert to pandas-on-Spark DataFrame
  - otherwise convert to Spark DataFrame
- if df is pandas-on-Spark DataFrame, convert to Spark DataFrame

Reply:
Would that introduce **two** different conversions (rather than one) in the case that the input is a pandas-on-Spark DataFrame? Wouldn't we want only a single conversion if possible?

Reply:
I think that's fine as converting from pandas-on-Spark to Spark DataFrame has no overhead.

```py
if isinstance(df, pd.DataFrame):
    if ps_available:
        df = ps.DataFrame(df)
    else:
        df = spark.createDataFrame(df)

if ps_available and isinstance(df, ps.DataFrame):
    df = df.to_spark()
```

Reply:
Thank you for this code example @ueshin !

My impression is that there's only 3 possibilities we care about for the original type of the df:
1. Spark DataFrame (no conversion needed)
1. pandas DataFrame
1. pandas-on-Spark DataFrame

Only the latter two need to be converted to Spark DataFrame.
Would the following flow work?
- if df is a Spark DataFrame, do nothing
- else if `pandas` is available and df is a pandas DataFrame, convert to Spark DataFrame
- else if `pyspark.pandas` is available and df is a pandas-on-Spark DataFrame, convert to Spark DataFrame
- else raise a helpful error message

My understanding is that we only need to import `pandas` and `pyspark.pandas` so that we can use `isinstance()`. Without those imports, then we can't determine if the df has those types using `isinstance()`.

An alternative is something like `type(df) == "pandas.core.frame.DataFrame"`, which has drawbacks like not supporting inheritance.

Reply:
The reason I want it to try to convert pandas DataFrame to pandas-on-Spark DataFrame first is that `ps.DataFrame(pandas_df)` knows how to convert pandas DataFrame better than `spark.createDataFrame(pandas_df)`.

Let's keep using `isinstance()`; otherwise we would miss something that inherits pandas/pandas-on-Spark DataFrame.

Reply:
I vastly underestimated the nuances of this PR! Thank you for your patient explanations 🙏 

Incorporating all your feedback here so that we can get all the details right as far upstream as possible:
https://github.com/dbt-labs/dbt-spark/pull/469

In hindsight, I should have created just a single PR there and invited you to be reviewer on it.

Reply:
Now that dbt-labs/dbt-spark#469 looks good to me, we can remove the macro in our side once it gets merged.